// truthy.js/target/lib/boolean.js
(function() {
  var boolean;

  boolean = {
    arrayey: function(a) {
      return this.existy(a) && Array.isArray(a);
    },
    booleany: function(a) {
      return this.existy(a) && typeof a === 'boolean';
    },
    existy: function(a) {
      return a != null;
    },
    extensibley: function(a) {
      return this.objecty(a) && Object.isExtensible(a);
    },
    finitey: function(a) {
      return (this.numbery(a) && isFinite(a)) || (this.stringy(a) && this.lengthy(a) && isFinite(a));
    },
    frozeny: function(a) {
      return this.objecty(a) && Object.isFrozen(a);
    },
    functiony: function(a) {
      return this.existy(a) && typeof a === 'function';
    },
    lengthy: function(a) {
      return this.existy(a) && ((a.hasOwnProperty('length') && a.length > 0) || (this.objecty(a) && Object.keys(a).length > 0));
    },
    nany: function(a) {
      return isNaN(a);
    },
    numbery: function(a) {
      return this.existy(a) && typeof a === 'number';
    },
    objecty: function(a) {
      return this.existy(a) && typeof a === 'object';
    },
    sealedy: function(a) {
      return this.objecty(a) && Object.isSealed(a);
    },
    stringy: function(a) {
      return this.existy(a) && typeof a === 'string';
    },
    truthy: function(a) {
      return this.existy(a) && (a !== false);
    }
  };

  module.exports = Object.freeze(boolean);

}).call(this);
